# AT4_Part_1
Part 1: Importing files, data wrangling, mathematical operations, plots and saving code on GitHub
This is part 1 of the Applied Bioinformatics Assessment Task 4. The analysis began with a gene expression TSV file, which was first read and processed. The mean value for each gene (row) was calculated, and a new column containing these mean values was added to the dataset. From this, the ten genes with the highest mean values were identified, and those with mean values less than 10 were also selected. A histogram of the mean values was then generated to visualize their distribution.

The next section involved analysing tree circumference data collected from two different sites between 2005 and 2020. The mean and standard deviation of tree circumference at both sites were determined for the start and end of the study period. Boxplots were created to compare the circumference distributions at the beginning and end for both sites. Finally, the mean difference and p-value between the two sites were calculated to assess any significant variation. 



# AT4_Part_2 
In Applied Bioinformatics Assessment Task 4, Part 2, the objective of this analysis was to compare the genomic data of Helicobacter anseris and Escherichia coli. The coding sequences (CDS) of both species were first downloaded, followed by a comparative analysis of their sequence lengths. In the next step, the total coding DNA was calculated and compared between the two organisms. Subsequently, the lengths of the coding sequences were analysed to determine the mean and median values, and the results were visualized using a boxplot.

In the fourth stage, the nucleotide and amino acid compositions of both species were examined and compared based on their respective quantities. Following this, codon usage bias was calculated and evaluated for similarities and differences between the two species. Finally, the analysis identified 10 protein k-mers for each organism, which were assessed to determine whether they were over-represented or under-represented.
